File: 6_zigzag_conversion/zigzag_conversion.py


class Solution:
    def convert(self, s, numRows):
        # 0.
        if numRows <= 1:
            return s

        # 1.
        lines = [""] * numRows
        row = 0
        step = 1

        # 2.
        for char in s:

            # 3.
            lines[row] += char

            # 4.
            if row == 0:
                step = 1
            elif row == numRows-1:
                step = -1

            # 5.
            row += step

        # 6.
        return "".join(lines)

File: 6_zigzag_conversion/zigzag_conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function (s, numRows) {
  // 0.
  if (numRows <= 1) {
    return s;
  }

  // 1.
  var lines = [];
  for (let i = 0; i < numRows; i++) lines.push("");
  // for (let i = 0; i < numRows; i++) lines[i] = [];
  var row = 0;
  var step = 1;

  // 2.
  for (let i = 0; i < s.length; i++) {
    // 3.
    lines[row] += s[i];

    // 4.
    if (row == 0 || row == numRows - 1) {
      step = step * -1;
    }

    // 5.
    row += step;
  }
  // 6.
  return lines.join("");
};
